File: src/php/PhantasyIslandPhpRemoteControl/Uav.php
<?php

namespace PhantasyIslandPhpRemoteControl;

/**
 * 此类是到 UAV 库的适配器，是对AirplaneManager的wrapper
 */
class UAV
{
    /** @var AirplaneManager */
    public $airs;

    public function __construct()
    {
        $this->airs = AirplaneManager::getInstance();
        $this->airs->flush();
        $this->airs->start();
        $this->airs->flush();
    }

    public function sleep($time)
    {
        $this->airs->sleep($time);
    }

    public function destroy()
    {
        // 销毁逻辑
    }

    public function add_uav($port)
    {
        return $this->airs->get_airplane($port);
    }

    public function p($port)
    {
        return $this->airs->get_airplane($port);
    }

    public function land($port)
    {
        $airplane = $this->p($port);
        if ($airplane) $airplane->land();
    }

    public function emergency($port)
    {
        $airplane = $this->p($port);
        if ($airplane) $airplane->stop();
    }

    public function takeoff($port, $high)
    {
        $airplane = $this->p($port);
        if ($airplane) $airplane->takeoff($high);
    }

    public function up($port, $distance)
    {
        $airplane = $this->p($port);
        if ($airplane) $airplane->up($distance);
    }

    public function down($port, $distance)
    {
        $airplane = $this->p($port);
        if ($airplane) $airplane->down($distance);
    }

    public function forward($port, $distance)
    {
        $airplane = $this->p($port);
        if ($airplane) $airplane->forward($distance);
    }

    public function back($port, $distance)
    {
        $airplane = $this->p($port);
        if ($airplane) $airplane->back($distance);
    }

    public function left($port, $distance)
    {
        $airplane = $this->p($port);
        if ($airplane) $airplane->left($distance);
    }

    public function right($port, $distance)
    {
        $airplane = $this->p($port);
        if ($airplane) $airplane->right($distance);
    }

    public function goto($port, $x, $y, $h)
    {
        $airplane = $this->p($port);
        if ($airplane) $airplane->goto($x, $y, $h);
    }

    public function flip($port, $direction)
    {
        $airplane = $this->p($port);
        if (!$airplane) return;
        
        switch ($direction) {
            case 'f': $airplane->flip_forward(); break;
            case 'b': $airplane->flip_back(); break;
            case 'r': $airplane->flip_right(); break;
            case 'l': $airplane->flip_left(); break;
        }
    }

    public function rotate($port, $degree)
    {
        $airplane = $this->p($port);
        if ($airplane) $airplane->rotate($degree);
    }

    public function cw($port, $degree)
    {
        $airplane = $this->p($port);
        if ($airplane) $airplane->cw($degree);
    }

    public function ccw($port, $degree)
    {
        $airplane = $this->p($port);
        if ($airplane) $airplane->ccw($degree);
    }

    public function speed($port, $speed)
    {
        $airplane = $this->p($port);
        if ($airplane) $airplane->speed($speed);
    }

    public function high($port, $high)
    {
        $airplane = $this->p($port);
        if ($airplane) $airplane->high($high);
    }

    public function led($port, $r, $g, $b)
    {
        $airplane = $this->p($port);
        if ($airplane) $airplane->led($r, $g, $b);
    }

    public function bln($port, $r, $g, $b)
    {
        $airplane = $this->p($port);
        if ($airplane) $airplane->bln($r, $g, $b);
    }

    public function rainbow($port, $r, $g, $b)
    {
        $airplane = $this->p($port);
        if ($airplane) $airplane->rainbow($r, $g, $b);
    }

    public function mode($port, $mode)
    {
        $airplane = $this->p($port);
        if ($airplane) $airplane->airplane_mode($mode);
    }

    public function stop($port)
    {
        $airplane = $this->p($port);
        if ($airplane) $airplane->stop();
    }

    public function hover($port)
    {
        $airplane = $this->p($port);
        if ($airplane) $airplane->hover();
    }
}

File: src/el/uav.erl
-module(uav).
-export([init/0, sleep/1, add_uav/1, land/1, emergency/1, takeoff/2, up/2, down/2, forward/2, back/2, left/2, right/2, goto/4, flip/2, rotate/2, cw/2, ccw/2, speed/2, high/2, led/4, bln/4, rainbow/4, mode/2, stop/1, hover/1]).

init() ->
    airplane_manager:flush(),
    airplane_manager:start(),
    airplane_manager:flush().

sleep(Time) ->
    airplane_manager:sleep(Time).

add_uav(Port) ->
    airplane_manager:get_airplane(Port).

p(Port) ->
    airplane_manager:get_airplane(Port).

land(Port) -> control_command:land(p(Port)).
emergency(Port) -> control_command:stop(p(Port)).
takeoff(Port, High) -> control_command:takeoff(p(Port), High).
up(Port, Dist) -> control_command:up(p(Port), Dist).
down(Port, Dist) -> control_command:down(p(Port), Dist).
forward(Port, Dist) -> control_command:forward(p(Port), Dist).
back(Port, Dist) -> control_command:back(p(Port), Dist).
left(Port, Dist) -> control_command:left(p(Port), Dist).
right(Port, Dist) -> control_command:right(p(Port), Dist).
goto(Port, X, Y, H) -> control_command:goto(p(Port), X, Y, H).
flip(Port, Dir) -> control_command:flip(p(Port), Dir).
rotate(Port, Deg) -> control_command:rotate(p(Port), Deg).
cw(Port, Deg) -> control_command:cw(p(Port), Deg).
ccw(Port, Deg) -> control_command:ccw(p(Port), Deg).
speed(Port, S) -> control_command:speed(p(Port), S).
high(Port, H) -> control_command:high(p(Port), H).
led(Port, R, G, B) -> control_command:led(p(Port), R, G, B).
bln(Port, R, G, B) -> control_command:bln(p(Port), R, G, B).
rainbow(Port, R, G, B) -> control_command:rainbow(p(Port), R, G, B).
mode(Port, M) -> control_command:mode(p(Port), M).
stop(Port) -> control_command:stop(p(Port)).
hover(Port) -> control_command:hover(p(Port)).
